### PR TITLE
document: show informational doc message after first reload

### PR DESCRIPTION
**Pull request by kugel-**
6f5d5db and d6d4728 disabled "maintain history on reload" by default,
with the intention to reenable it when we have a better method to
make it discoverable for the user. This was necessary since it became
enabled by default but could be surprising given Geany warned about
losing data before.

This commit tries to resolve the discoverability, by providing an
informational doc message that is shown once to the user, after the first
reload. The doc message also gives the option to disable this feature.

### CODE REVIEW

File: src/document.c [hunk 1]
```diff
@@ -1520,6 +1520,24 @@ void document_open_files(const GSList *filenames, gboolean readonly, GeanyFilety
 }
 
 
+static void on_keep_edit_history_on_reload_response(GtkWidget *bar, gint response_id, GeanyDocument *doc)
+{
+	if (response_id == GTK_RESPONSE_NO)
+	{
+		file_prefs.keep_edit_history_on_reload = FALSE;
+		document_reload_force(doc, doc->encoding);
```
Comment (b4n) [inline src/document.c]: It probably would be great if we could drop the history without actually reloading a second time (`sci_empty_undo_buffer`/`document_undo_clear`, but I'm not 100% sure why/if we don't have a unified way)

File: src/document.c [hunk 2]
```diff
@@ -1545,6 +1564,21 @@ gboolean document_reload_force(GeanyDocument *doc, const gchar *forced_enc)
...
+	{
+		bar = document_show_message(doc, GTK_MESSAGE_INFO,
+						on_keep_edit_history_on_reload_response,
+						GTK_STOCK_OK, GTK_RESPONSE_ACCEPT,
+						"Discard history", GTK_RESPONSE_NO,
+						NULL, 0, "You can revert the buffer to the previous state "
+						"simply by undoing unless you choose to discard the history. "
+						"This message will not be displayed again. "
+						"Your choice is can be changed in the various preferences.",
+						"The file has been reloaded.");
```
Comment (b4n) [inline src/document.c]: These strings need to be translatable

Comment (b4n) [inline src/document.c]: These messages probably need to be reviewed by @elextr

File: src/keyfile.c [hunk 3]
```diff
@@ -235,6 +235,8 @@ static void init_pref_groups(void)
 		"use_gio_unsafe_file_saving", TRUE);
 	stash_group_add_boolean(group, &file_prefs.keep_edit_history_on_reload,
 		"keep_edit_history_on_reload_125", FALSE);
```
Comment (b4n) [inline src/keyfile.c]: This needs to be renamed back to `keep_edit_history_on_reload` without the `_125` suffix to reset the default value.

File: src/document.c (hunk 1 above)
Comment (kugel-) [inline src/document.c]: This way it's a one-liner, and the whole infomsg is a one-timer. So it's not really worth any optimization.

**Comment by b4n** [comment]
I would have tried to do it differently with a checkbox to ask whether to remember the choice, but your version seems reasonable and easy to implement.
I'll have to try it for real, but it looks nice.

File: src/document.c (hunk 2 above)
Comment (kugel-) [inline src/document.c]: Argh yes, that was the whole point of not doing it for 1.25.

File: src/keyfile.c (hunk 3 above)
Comment (kugel-) [inline src/keyfile.c]: Right, I saved that for a separate commit.

File: src/document.c [hunk 4]
```diff
@@ -1545,6 +1564,21 @@ gboolean document_reload_force(GeanyDocument *doc, const gchar *forced_enc)
 	pos = sci_get_current_position(doc->editor->sci);
 	new_doc = document_open_file_full(doc, NULL, pos, doc->readonly, doc->file_type, forced_enc);
 
+	if (file_prefs.keep_edit_history_on_reload && file_prefs.show_keep_edit_history_on_reload_msg)
+	{
+		bar = document_show_message(doc, GTK_MESSAGE_INFO,
+						on_keep_edit_history_on_reload_response,
+						GTK_STOCK_OK, GTK_RESPONSE_ACCEPT,
+						"Discard history", GTK_RESPONSE_NO,
+						NULL, 0, "You can revert the buffer to the previous state "
```
Comment (elextr) [inline src/document.c]: I would be explicit "You can revert the buffer to the state before the reload"

File: src/document.c [hunk 5]
```diff
@@ -1545,6 +1564,21 @@ gboolean document_reload_force(GeanyDocument *doc, const gchar *forced_enc)
...
+						"simply by undoing unless you choose to discard the history. "
+						"This message will not be displayed again. "
```
Comment (elextr) [inline src/document.c]: It is traditional to have the won't repeat line last"

File: src/document.c [hunk 6]
```diff
@@ -1545,6 +1564,21 @@ gboolean document_reload_force(GeanyDocument *doc, const gchar *forced_enc)
...
+						"Your choice is can be changed in the various preferences.",
```
Comment (elextr) [inline src/document.c]: remove the "is"

File: src/document.c (hunk 5 above)
Comment (kugel-) [inline src/document.c]: Pardon?

File: src/document.c (hunk 4 above)
Comment (kugel-) [inline src/document.c]: The message is already rather long. I'd prefer to not make it even longer without good reason.

**Comment by elextr** [comment]
In general LGTM, thanks for remembering to do this.

File: src/document.c (hunk 4 above)
Comment (elextr) [inline src/document.c]: It is one or two characters longer, but *much* more explicit.

Comment (kugel-) [inline src/document.c]: I disagree but well, I'll change it.

But remember that this line appears in the GUI directly under the headline "The file has been reloaded." so it's sufficiently clear IMO.

**Comment by codebrainz** [comment]
On 15-10-02 03:31 PM, Colomban Wendling wrote:
> I would have tried to do it differently with a checkbox to ask whether to remember the choice, but your version seems reasonable and easy to implement.
> I'll have to try it for real, but it looks nice.
>

What does it use if not a checkbox?

It can't just show once and never again since the message can be 
replaced with a higher priority one (IIUC), the user might not even see 
it the first time, or even if they do see it, they might just want to 
dismiss it, to deal with it next time.

**Comment by kugel-** [comment]
>
>What does it use if not a checkbox?
>
>It can't just show once and never again since the message can be 
>replaced with a higher priority one (IIUC), the user might not even see
>
>it the first time, or even if they do see it, they might just want to 
>dismiss it, to deal with it next time.
>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/672#issuecomment-145272642

It actually shows the message only once, *except* if it was replaced by a higher priority one.

So it's only shown once if the user clicked it away.

**Comment by codebrainz** [comment]
> It actually shows the message only once, *except* if it was replaced by a higher priority one. So it's only shown once if the user clicked it away.

Ok that makes more sense. What happens if the message is replaced, does it discard history or preserve it (until the next time the message is shown)?

**Comment by kugel-** [comment]
The history is kept in this case (that's the default anyway)

**Comment by elextr** [comment]
I just realised that if a user ignores the info and does something else before selecting to delete history it deletes the undo for the extra things the user did.

As this is a once only notification/question, it probably should be a modal dialog to prevent any further actions until it is actioned.

**Comment by kugel-** [comment]
Err, no, modal is even worse for (mostly) informational messages.

**Comment by elextr** [comment]
It isn't an asynchronous message, it is part of the reload action.

1. the user presses reload the first time in 1.26

2. a modal dialog comes up noting the new behaviour and offering to keep/delete history

3. user makes a selection and the dialog goes away forever and the reload completes

**Comment by kugel-** [comment]
@elextr @b4n @codebrainz @frlan I'm sorry I won't manage to drive this forward until after my return from vacations. Since the string freeze is nearing again I ask that one of you pick this up and finish this up.

Or at least define a message string so that this doesn't fail again to be merged due to string freeze.

**Comment by kugel-** [comment]
@elextr @b4n @codebrainz @frlan string freeze on weekend, we gotta hurry.

**Comment by codebrainz** [comment]
I personally don't like this feature nor think it should be enabled by default, and since it's one more thing I'll have to change when installing Geany, I'm certainly in no hurry myself to get it merged :)

**Comment by elextr** [comment]
@kugel- you havn't changed the message both for the change you agreed to (even if you were unsure about) and for the plain English error (unneeded "is") and you havn't added the extra commit to rename the pref.

So its you that needs to hurry :)

**Comment by kugel-** [comment]
Besides that I asked someone to pick it up before my vacations, I don't feel we have worked out the final message and ui yet, therefore I did no change at all yet.

**Comment by kugel-** [comment]
Updated. I think I could improve the message further, I hope you agree

**Comment by elextr** [comment]
I think you mean that you *did* improve the message, and I agree thats ok.

Unless b4n objects or does it himself I'll probably commit it tomorrow.

File: src/document.c [hunk 7]
```diff
@@ -1545,6 +1564,21 @@ gboolean document_reload_force(GeanyDocument *doc, const gchar *forced_enc)
...
+	if (file_prefs.keep_edit_history_on_reload && file_prefs.show_keep_edit_history_on_reload_msg)
+	{
+		bar = document_show_message(doc, GTK_MESSAGE_INFO,
+						on_keep_edit_history_on_reload_response,
+						GTK_STOCK_OK, GTK_RESPONSE_ACCEPT,
+						_("Discard history"), GTK_RESPONSE_NO,
+						NULL, 0, _("The buffer's previous state is stored in the history. "
+						"Undoing restores it. You can disable this by discarding the history upon "
+						"reload. This message will not be displayed again. "
+						"Your choice can be changed in the various preferences."),
```
Comment (b4n) [inline src/document.c]: @elextr What about

> The buffer's previous state is stored in the undo history, and then file reloading can be undone.  You can disable this by discarding the history upon reload.  This message will not be displayed again, but your can change the behavior later in the various preference.

While I generally like short sentences, I feel like the 5 sentences here are a bit heavy to read.

**Comment by b4n** [comment]
Documentation changes from https://github.com/geany/geany/commit/6f5d5db2cb6da4970aa78a64efee3a91727fea51 have to also be reverted to match actual behavior

File: src/document.c (hunk 7 above)
Comment (kugel-) [inline src/document.c]: This becomes longer again. It's already unwidely long IMO.

If the number of sentences is the problem why not just join the sentences? "The buffer's previous state is stored in the history and undoing restores it. You can disable this by discarding the history upon reload. This message will not be displayed again but your choice can be changed in the various preferences."

Comment (b4n) [inline src/document.c]: Well, okay, although I don't really see much difference in length, it sounds good enough your way just removing the two stops.

Comment (codebrainz) [inline src/document.c]: Might be useful to add the name of the various preference so user doesn't have to hunt for it or try/test different preferences.

**Comment by codebrainz** [comment]
We should probably also add a note in the docs warning about using this option with large files, since it will "leak" any reloaded documents for the entire document's lifetime. I can just imagine someone filing a bug about "Geany leaks hundreds of megabytes when I reload many large files after changed on disk" or stuff like that.

Are we definitively sure this should be enabled by default? (I'm not :)

**Comment by kugel-** [comment]
@b4n fixed your remarks

**Comment by b4n** [comment]
> We should probably also add a note in the docs warning about using this option with large files, since it will "leak" any reloaded documents for the entire document's lifetime. I can just imagine someone filing a bug about "Geany leaks hundreds of megabytes when I reload many large files after changed on disk" or stuff like that.
>
> Are we definitively sure this should be enabled by default? (I'm not :)

Well… I think it's a very nice feature; but OTOH you're right that until (unless?) Scintilla starts to store a diff in the history instead of the whole change (or we do), it'll use a largish amount of memory.  And it seems very possible most people tend to reload mostly largish files (logs, DB dumps or whatnot generated outside and getting updated), so it might have a larger impact than on other features.
However, closing the document releases the memory, so it "only" impacts while the document is in use.

I guess ideally we'd reduce the memory impact the feature has, but until then I'm not 100% certain.

**Comment by b4n** [comment]
Otherwise LGTM

**Comment by kugel-** [comment]
The feature was initially accepted with default = on, so unless there are actual problems I don't see an evidence to keep the off default.

I think it's a nice and convinient feature which should be on ultimately on by default. its memory consumption is a detail and can be fixed elsewhere (if it's even necessary)

**Comment by kugel-** [comment]
String freeze is tomorrow so please merge

**Comment by elextr** [comment]
@b4n, don't punish the innocent (those who will use the feature as intended) just because somebody *might* use it wrong.

**Comment by b4n** [comment]
@elextr fair enough for me.  Could you please review the string one last time and merge this?
Don't forget to close #562 in the merge commit

**Comment by codebrainz** [comment]
> [...] so unless there are actual problems I don't see an evidence to keep the off default.

I mentioned a number of issues with it in the mailing list thread linked from the PR you made to disable it by default.

I don't think we should cram this in just because of string freeze, especially with the open issues, the UI issues, and that we will be releasing again in a few months. I'm not opposed to the feature, I'm opposed to the implementation (re-using an existing feature to do something else, something unexpected of the existing feature, as well as the infobar without a checkbox that can be hidden from the user without ever seeing it).

**Comment by codebrainz** [comment]
Bummer ... I think we should reconsider enabling this feature by default, or at the very least fix the issues with it.

**Comment by kugel-** [comment]
Am 31. Oktober 2015 02:52:01 MEZ, schrieb Matthew Brush <notifications@github.com>:
>> [...] so unless there are actual problems I don't see an evidence to
>keep the off default.
>
>I mentioned a number of issues with it in the mailing list thread
>linked from the PR you made to disable it by default.
>
>I don't think we should cram this in just because of string freeze,
>especially with the open issues, the UI issues, and that we will be
>releasing again in a few months. I'm not opposed to the feature, I'm
>opposed to the implementation (re-using an existing feature to do
>something else, something unexpected of the existing feature, as well
>as the infobar without a checkbox that can be hidden from the user
>without ever seeing it).
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/672#issuecomment-152685983

Carrying this discussion to the ML.

Can you elaborate on the issues you see?

As to the UI. It's important that we get any UI for 1.26, therefore we had to manage before string freeze. We can introduce a better UI fir 1.27. I'm not opposed to a check box but would be more effort. And it would presumably checked by default anyway, no?

**Pull request merged**